### PR TITLE
Add the optional health-port flag used for the health check URL

### PR DESCRIPTION
**Pull request by secat**
related to #25 

I have tested locally with the following command:
```
./sidekick --health-path=/ --health-port=13133 --address :8000 http://localhost:8888
```

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -191,9 +196,20 @@ func registerMetricsRouter(router *mux.Router) error {
 	return nil
 }
 
+// getHealthCheckURL - extracts the health check URL.
+func getHealthCheckURL(endpoint, healthCheckPath string, healthCheckPort int) string {
+	healthCheckURL, _ := url.Parse(strings.TrimSuffix(endpoint, slashSeparator) + healthCheckPath)
+	if healthCheckPort != 0 {
+		host, _, _ := net.SplitHostPort(healthCheckURL.Host)
+		healthCheckURL.Host = host + colonSeparator + strconv.Itoa(healthCheckPort)
```
Comment (harshavardhana) [inline main.go:204]: You should use net.JoinHostPort()

File: main.go [hunk 2]
```diff
@@ -191,9 +196,20 @@ func registerMetricsRouter(router *mux.Router) error {
 	return nil
 }
 
+// getHealthCheckURL - extracts the health check URL.
+func getHealthCheckURL(endpoint, healthCheckPath string, healthCheckPort int) string {
+	healthCheckURL, _ := url.Parse(strings.TrimSuffix(endpoint, slashSeparator) + healthCheckPath)
+	if healthCheckPort != 0 {
```
Comment (harshavardhana) [inline main.go:202]: This function can return early when healthCheckPort is not set without resorting to url.Parse, passing port is rare so we should keep the code as such.

**Comment by secat** [comment]
@harshavardhana I have also implemented a unit test for the `getHealthCheckURL` function.

Do you want me to push the unit test file in this pull request?

Here is the content of the `main_test.go` that I can push if desired:
```golang
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func Test(t *testing.T) {
	// ----- conditions -------------------------------------------------------
	testCases := []struct {
		name                   string
		endpoint               string
		healthCheckPath        string
		healthCheckPort        int
		expectedHealthCheckURL string
	}{
		{
			name:                   "PortSetToZero",
			endpoint:               "http://minio1:9000",
			healthCheckPath:        "/minio/health/ready",
			expectedHealthCheckURL: "http://minio1:9000/minio/health/ready",
		},
		{
			name:                   "PortSetToNonZeroValue",
			endpoint:               "http://minio1:9000",
			healthCheckPath:        "/minio/health/ready",
			healthCheckPort:        4242,
			expectedHealthCheckURL: "http://minio1:4242/minio/health/ready",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			// ----- call -----------------------------------------------------
			healthCheckURL := getHealthCheckURL(tc.endpoint, tc.healthCheckPath, tc.healthCheckPort)

			// ----- verify ---------------------------------------------------
			assert.Equal(t, tc.expectedHealthCheckURL, healthCheckURL)
		})
	}
}
```

**Comment by harshavardhana** [comment]
> @harshavardhana I have also implemented a unit test for the `getHealthCheckURL` function.
> 
> Do you want me to push the unit test file in this pull request?
> 
> Here is the content of the `main_test.go` that I can push if desired:
> ```golang
> package main
> 
> import (
> 	"testing"
> 
> 	"github.com/stretchr/testify/assert"
> )
> 
> func Test(t *testing.T) {
> 	// ----- conditions -------------------------------------------------------
> 	testCases := []struct {
> 		name                   string
> 		endpoint               string
> 		healthCheckPath        string
> 		healthCheckPort        int
> 		expectedHealthCheckURL string
> 	}{
> 		{
> 			name:                   "PortSetToZero",
> 			endpoint:               "http://minio1:9000",
> 			healthCheckPath:        "/minio/health/ready",
> 			expectedHealthCheckURL: "http://minio1:9000/minio/health/ready",
> 		},
> 		{
> 			name:                   "PortSetToNonZeroValue",
> 			endpoint:               "http://minio1:9000",
> 			healthCheckPath:        "/minio/health/ready",
> 			healthCheckPort:        4242,
> 			expectedHealthCheckURL: "http://minio1:4242/minio/health/ready",
> 		},
> 	}
> 
> 	for _, tc := range testCases {
> 		t.Run(tc.name, func(t *testing.T) {
> 			// ----- call -----------------------------------------------------
> 			healthCheckURL := getHealthCheckURL(tc.endpoint, tc.healthCheckPath, tc.healthCheckPort)
> 
> 			// ----- verify ---------------------------------------------------
> 			assert.Equal(t, tc.expectedHealthCheckURL, healthCheckURL)
> 		})
> 	}
> }
> ```

No testify stuff please just use regular Go testing if you want to add tests.

File: main.go (hunk 1 above)
Comment (secat) [reply main.go]: Updated the code using `net.JoinHostPort()`

File: main.go (hunk 2 above)
Comment (secat) [reply main.go]: Updated the code using a fast path when the `healthCheckPort` is set to `0`

**Comment by secat** [comment]
@harshavardhana since this is my first contribution for this project, I was wondering how it was working for review comments: when a comment is addressed, should I click the `resolve conversation` button, or should it be done by the reviewer?

**Comment by harshavardhana** [comment]
> @harshavardhana since this is my first contribution for this project, I was wondering how it was working for review comments: when a comment is addressed, should I click the `resolve conversation` button, or should it be done by the reviewer?

You don't have to, it is upto you. It can be done either by you or by the reviewer. Generally if the fix in the place where the reviewer commented. GitHub neatly highlights that the review was addressed.

File: main.go [hunk 3]
```diff
@@ -191,9 +193,23 @@ func registerMetricsRouter(router *mux.Router) error {
...
 
+// getHealthCheckURL - extracts the health check URL.
+func getHealthCheckURL(endpoint, healthCheckPath string, healthCheckPort int) string {
+	healthCheckURL, _ := url.Parse(strings.TrimSuffix(endpoint, slashSeparator) + healthCheckPath)
+	if healthCheckPort == 0 {
+		return healthCheckURL.String()
+	}
+
+	// Set health check port
+	host, _, _ := net.SplitHostPort(healthCheckURL.Host)
```
Comment (harshavardhana) [inline main.go:204]: you can use URL helper like 

```
healthCheckURL.Host  = net.JoinHostPort(healthCheckURL.Hostname(), strconv.Itoa(healthCheckPort))
```

Comment (harshavardhana) [reply main.go:204]: Also checks if healthCheckPort > 65535 or < 0

File: main_test.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package main
```
Comment (harshavardhana) [inline main_test.go:1]: Missing license header to be added from main.go

File: main_test.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+			healthCheckPath: "/minio/health/ready",
+			want:            "http://minio1:9000/minio/health/ready",
+		},
+		{
+			name:            "PortSetToNonZeroValue",
+			endpoint:        "http://minio1:9000",
+			healthCheckPath: "/minio/health/ready",
+			healthCheckPort: 4242,
+			want:            "http://minio1:4242/minio/health/ready",
+		},
```
Comment (harshavardhana) [inline main_test.go:28]: You could add test cases for negative aspects as well such as incorrect healthCheckPort and also incorrect endpoint - incase of latter perhaps the code might crash.

Comment (secat) [reply main_test.go:43]: Separated tests for valid and invalid use cases.
